### ISSUE TITLE
Configure message send logic in the network server

### ISSUE THREAD

**Issue by fyrchik** [nspcc-dev/neo-go#2678]
Currently we send any message to a 2/3 of our peers, which increases performance based on our benchmarks.
However, there is a number of ways to improve this:
1. If there are _only_ 4 nodes in the system (privnet), consensus messages can be lost easily (3 nodes decided to sent a message to each other, 4th node receives nothing). This leads to a ChangeView which can be undesireable.
2. If we have lots of peers in a big network, 2/3 of peers can be even more that we really need.
3. If we received an inventory item from a peer, it would be nice to not sent it back an INV message. Here we don't need to change config at all and it makes situation (1) much less probable (impossible?). If this can be done with little overhead, I also expect this to improve TPS.

E.g. we could have `BroadcastThreshold: 3` to make sure we send a message to at least 3 of our peers. This can also make it easier to benchmark in different settings.

**Comment by roman-khimov** [comment]
Hi, #608. I really don't like any settings for this, but some reasonable per-peer LRU cache for №3 can be tried, I think. At least to see how it works and then we can decide on whether it's the right way to go or we have to do some magic settings.

**Comment by roman-khimov** [comment]
1. Changing to "send-to-all" model by default means degrading performance (it's 100% known), we can't have a regression like this.
2. Making it an option raises the question of this option status and maintenance.
3. 0.99.3 is not the right release for changes like this, we have a plan for it and we better stick to it.
4. Changing view is a _normal_ operation, there is nothing inherently bad about it. If any upper layers have problems with this, they need to be fixed ASAP. If the behavior questioned here makes CVs more probable then it's good, you can see upper layer problems easier in sterile environment.

**Comment by realloc** [comment]
I'd suggest adding `BroadcastThreshold` option and make it obsolete when gossip improvements appear.

**Comment by roman-khimov** [comment]
It will degrade performance in your particular scenario. You may have worse situation after the change than you have now. And it's not that you have any big problem at all, the chain works fine. It's not broken, so it needs no fix, especially any urgent one.

**Comment by roman-khimov** [comment]
Let's start with #2738, I'll try reworking queue handling code a bit tomorrow before we move to the inv hash cache (it's drafted in the 2d463670ab69ba4cc840c5aec88146b50753ce3a and there are some interesting results for it if a big enough cache is used (4096 is not big enough), but that's the next thing.

**Comment by roman-khimov** [comment]
On the per-peer inventory hash cache and checks against it before sending any inventories. While the idea seems to be very reasonable and theoretically important even for 4+1 neo-bench case (where RPC node is the source of transactions and CNs may never send tx inventories to it, it's a clear waste of time), it doesn't help in practice.

I've tried a simple LRU hash cache for it and then memory-optimized one, they've shown some improvements after #2738, but after a rebase on top of #2741 it either doesn't change anything or directly degrades (10-20%%) performance. This is a somewhat unexpected, but reliably reproducible behavior. Keeping this cache also has some substantial memory cost to it, it has to be around mempool size number of 32 byte elements per-peer, for the simple LRU cache that's around 3MB per-peer. I've tried 8-byte hash parts, which is better memory-wise, but this doesn't change anything wrt performance, it just works bad.

So we won't do do hash caching and any associated logic.

One thing I'd like to try is solving #608 with `int(2.5 * log(number of nodes) + 0.5)` formula. The number of nodes can be derived from connected+unconnected addresses known to the discovery service. In some cases it'll be perfect (like 4-node setup, all addresses are known there), it some cases it won't (public networks), but to me it seems to be accurate enough to do this. Fan-out number is log-dependent on this estimation anyway, so if we see just a 100 nodes out of 200 real ones, it's not a big error (and we tend to not have networks with thousands of nodes). The resulting fan-out number of nodes will also be the same as the current 2/3 for low number of nodes, so the behavior will be the same there.

The only thing I'm not sure about is should there be any configuration for it or not. I don't like having a lot of knobs no one uses/knows how to use/potentially breaking the system, yet at the same time I can imagine some `GossipOptimization` coefficient (0-100) that will allow to slide between magic log formula above as the minimal fan-out value (100 optimization) and "send to everyone" as the maximum one (0 optimization). Do we need that, do we want that? :rocket: below for the option, :tada: for going without options or post any other opinions in comments. @fyrchik, @AnnaShaleva, @realloc, @alexvanin.

**Comment by roman-khimov** [comment]
Inventory hash cache results:
![cpu_four_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353625-0a3b8320-1b2f-4a28-8124-10b8d700da56.png)
![cpu_four_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353636-98613f3e-7aae-4422-b1ba-7581a586e505.png)
![cpu_seven_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353641-02348442-c844-4f22-b10b-70505dde4091.png)
![cpu_seven_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353643-03366d66-8c85-46b1-8d4d-6ecc75da33b3.png)
![mem_four_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353646-c4d39db7-df94-4ebd-ab0b-446fa5676376.png)
![mem_four_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353647-c497b5de-b131-476d-85bc-97db44205aac.png)
![mem_seven_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353650-e87aaece-117e-42e4-aca5-d2a0579c4c35.png)
![mem_seven_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353654-e0c8c199-d7de-4859-a437-097b3b2027cb.png)
![ms_per_block_four_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353655-0d3ccaa5-8d55-4e35-aa18-130e1ab905df.png)
![ms_per_block_four_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353657-07231678-a7ab-40c2-80a9-964a45583550.png)
![ms_per_block_seven_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353658-fbefccea-d212-431a-85cb-3ad69f6f100c.png)
![ms_per_block_seven_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353664-ac2d828e-cfff-4454-85ec-0dfe20875409.png)
![tpb_four_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353668-2aa397d0-111d-47d6-9469-ec04e18de06f.png)
![tpb_four_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353671-3d9ff85e-7e18-4479-9f8d-c4a06503bedf.png)
![tpb_seven_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353674-43f705b0-0890-44ef-9c52-67a56c0cf8a3.png)
![tpb_seven_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353677-598f9f8d-0bbd-4ae0-bce8-af728e2c73f6.png)
![tps_four_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353682-4f86a8ac-7d81-4b2c-9995-7121c3b0794a.png)
![tps_four_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353686-e39e435b-b2fd-4222-a8fe-1bc313acccc0.png)
![tps_seven_(30_wrk,_50K_pool,_200ms)](https://user-images.githubusercontent.com/22092804/195353692-204a3e13-a6bf-477b-818b-f4decc4d224c.png)
![tps_seven_(30_wrk,_50K_pool)](https://user-images.githubusercontent.com/22092804/195353696-25445630-05b0-4fb0-a856-097afa859b83.png)

**Comment by realloc** [comment]
For small clusters operating under production load, not benchmarks, it is better to lose some performance instead of losing some messages and having block propagation delays.
